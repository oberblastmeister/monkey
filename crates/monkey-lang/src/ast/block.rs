use crate::{ast, Parse, ParseError, ParseResult, Parser, Peek, Peeker, Spanned};

#[derive(Debug, Clone, PartialEq)]
pub struct Block {
    pub open_bracket: T!['{'],
    pub stmts: Vec<ast::Stmt>,
    pub close_bracket: T!['}'],
}

impl Parse for Block {
    fn parse(p: &mut Parser) -> ParseResult<Self> {
        let open_bracket = p.parse()?;

        let mut stmts = Vec::new();

        loop {
            stmts.push(p.parse()?);
            if p.is_eof()? {
                break;
            }
        }

        let close_bracket = p.parse()?;

        Ok(Block {
            open_bracket,
            stmts,
            close_bracket,
        })
    }
}
