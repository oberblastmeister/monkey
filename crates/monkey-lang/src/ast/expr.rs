use crate::{ast, Parse, ParseError, ParseResult, Parser, Peek, Peeker, Spanned};

#[derive(Debug, Clone, PartialEq)]
pub enum Expr {
    Binary(Box<ast::ExprBinary>),
    Lit(ast::Lit),
}

impl Parse for Expr {
    #[inline]
    fn parse(p: &mut Parser) -> ParseResult<Self> {
        expr(p)
    }
}

#[inline]
fn expr(p: &mut Parser) -> ParseResult<Expr> {
    expr_bp(p, 0)
}

fn expr_bp(p: &mut Parser, min_bp: u8) -> ParseResult<Expr> {
    let mut lhs = match p.nth(0)? {
        K!['('] => expr_paren(p)?,
        _ if p.peek::<ast::Lit>() => Expr::Lit(p.parse()?),
        t => panic!("Bad token: {:?}", t),
    };

    loop {
        if p.peek::<ExprEnd>() {
            break;
        };

        let op = ast::BinOp::from_peeker(&mut p.peeker).ok_or_else(|| {
            ParseError::expected(&p.next().unwrap(), "Expected a binary operator")
        })?;

        let (l_bp, r_bp) = op.binding_power();

        if l_bp < min_bp {
            break;
        }

        lhs = {
            let token = p.next()?;

            let rhs = expr_bp(p, r_bp)?;

            Expr::Binary(Box::new(ast::ExprBinary {
                lhs,
                op,
                rhs,
                token,
            }))
        };

        continue;
    }

    Ok(lhs)
}

struct ExprEnd;

impl Peek for ExprEnd {
    fn peek(p: &mut Peeker) -> bool {
        matches!(p.nth(0), K![eof] | K![;] | K![')'])
    }
}

fn expr_paren(p: &mut Parser) -> ParseResult<Expr> {
    assert_eq!(p.nth(0).unwrap(), K!['(']);

    p.next().unwrap();
    let lhs = expr_bp(p, 0)?;
    p.parse::<T![')']>()?;
    Ok(lhs)
}
