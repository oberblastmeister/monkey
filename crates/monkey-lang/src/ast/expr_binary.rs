use crate::{ast, Parse, ParseError, ParseResult, Parser, Peek, Peeker, Spanned};

#[derive(Debug, Clone, PartialEq)]
pub struct ExprBinary {
    pub lhs: ast::Expr,
    /// the token of the operator
    pub token: ast::Token,
    pub op: BinOp,
    pub rhs: ast::Expr,
}

impl ExprBinary {
    pub fn with_lhs(
        p: &mut Parser<'_>,
        lhs: ast::Expr,
        op: BinOp,
    ) -> ParseResult<ExprBinary> {
        assert!(BinOp::from_peeker(&mut p.peeker).is_some());

        let token = p.next()?;

        Ok(ExprBinary {
            lhs,
            token,
            op,
            rhs: p.parse()?,
        })
    }
}

#[derive(Debug, Clone, Copy, PartialEq)]
pub enum BinOp {
    Add,
    Sub,
    Mul,
    Div,
}

use BinOp::*;

impl BinOp {
    pub fn from_peeker(p: &mut Peeker<'_>) -> Option<BinOp> {
        Some(match p.nth(0) {
            K![+] => Add,
            K![-] => Sub,
            K![*] => Mul,
            K![/] => Div,
            _ => return None,
        })
    }

    pub fn binding_power(self) -> (u8, u8) {
        match self {
            Add | Sub => (1, 2),
            Mul | Div => (3, 4),
        }
    }
}

impl Peek for BinOp {
    fn peek(p: &mut Peeker) -> bool {
        BinOp::from_peeker(p).is_some()
    }
}
