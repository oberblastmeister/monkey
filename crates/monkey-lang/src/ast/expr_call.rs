use crate::{ast, Parse, ParseResult, Parser};

#[derive(Debug, Clone, PartialEq)]
pub struct ExprCall {
    ident: T![ident],
    arguments: Arguments,
}

impl Parse for ExprCall {
    fn parse(p: &mut Parser) -> ParseResult<Self> {
        Ok(ExprCall {
            ident: p.parse()?,
            arguments: p.parse()?,
        })
    }
}

#[derive(Debug, Clone, PartialEq)]
pub struct Arguments(ast::List<T![ident], T!['{'], T!['}']>);

impl Parse for Arguments {
    fn parse(p: &mut Parser) -> ParseResult<Self> {
        Ok(Arguments(p.parse()?))
    }
}
