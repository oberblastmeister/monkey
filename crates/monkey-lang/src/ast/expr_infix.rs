use crate::{ast, Parse, ParseError, ParseResult, Parser, Peek, Peeker, Spanned};

#[derive(Debug, Clone, PartialEq)]
pub struct ExprInfix {
    /// The left hand side of the expression
    pub lhs: ast::Expr,
    /// the token of the operator
    pub token: ast::Token,
    /// The operator type
    pub op: ast::Op,
    /// The right hand side of the expression
    pub rhs: ast::Expr,
}

expr_parse!(Infix, ExprInfix, "binary expression");
