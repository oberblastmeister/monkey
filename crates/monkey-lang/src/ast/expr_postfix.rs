use crate::{ast, Parse, ParseError, ParseResult, Parser, Peek, Peeker, Spanned};

#[derive(Debug, Clone, PartialEq)]
pub struct ExprPostfix {
    /// The right hand side of the expression
    pub lhs: ast::Expr,
    /// the token of the operator
    pub token: ast::Token,
    /// The operator type
    pub op: ast::Op,
}

expr_parse!(Postfix, ExprPostfix, "postfix expression");
