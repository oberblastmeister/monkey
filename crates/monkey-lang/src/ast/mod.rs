#[macro_use]
pub mod generated;

pub use generated::TokenKind;

macro_rules! expr_parse {
    ($ty:ident, $local:ty, $expected:literal) => {
        impl crate::Parse for $local {
            fn parse(p: &mut crate::Parser<'_>) -> Result<Self, crate::ParseError> {
                let span = p.tok_at(0)?.span;

                match crate::ast::Expr::parse(p)? {
                    crate::ast::Expr::$ty(expr) => Ok(*expr),
                    _ => Err(crate::ParseError::msg(&span, $expected)),
                }
            }
        }
    };
}

mod token;

mod expr;
mod expr_lit;
mod expr_infix;
mod expr_prefix;
mod expr_postfix;
mod expr_call;

mod lit;
mod lit_bool;
mod lit_num;
mod lit_str;
mod lit_fn;

mod stmt;
mod stmt_expr;
mod stmt_return;
mod stmt_let;
mod stmt_if;

mod block;

mod file;

mod op;

mod list;

pub use token::Token;

pub use expr::Expr;
pub use expr_lit::ExprLit;
pub use expr_infix::ExprInfix;
pub use expr_prefix::ExprPrefix;
pub use expr_postfix::ExprPostfix;
pub use expr_call::ExprCall;

pub use op::Op;

pub use lit::Lit;
pub use lit_bool::LitBool;
pub use lit_num::LitNum;
pub use lit_str::LitStr;
pub use lit_fn::LitFn;

pub use stmt::Stmt;
pub use stmt_expr::StmtExpr;
pub use stmt_return::StmtReturn;
pub use stmt_let::StmtLet;
pub use stmt_if::StmtIf;

pub use block::Block;

pub use file::File;

pub use list::List;
