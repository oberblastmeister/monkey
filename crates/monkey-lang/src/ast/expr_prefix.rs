use crate::ast;

#[derive(Debug, Clone, PartialEq)]
pub struct ExprPrefix {
    /// the token of the operator
    pub token: ast::Token,
    /// The operator type
    pub op: ast::Op,
    /// The right hand side of the expression
    pub rhs: ast::Expr,
}

expr_parse!(Prefix, ExprPrefix, "prefix expression");
