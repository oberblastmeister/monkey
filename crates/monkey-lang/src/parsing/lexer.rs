#![allow(clippy::unnecessary_wraps)]

use smol_str::SmolStr;
use unicode_xid::UnicodeXID;

use super::{ParseError, ParseResult, ToEof};
use crate::ast::{Token, TokenKind};
use crate::Span;
use std::str::Chars;

#[derive(Debug)]
struct Source<'a> {
    start: usize,
    input_len: usize,
    input: &'a str,
    chars: Chars<'a>,
}

impl<'a> Source<'a> {
    fn new(input: &'a str) -> Source<'a> {
        Source {
            start: 0,
            input,
            input_len: input.len(),
            chars: input.chars(),
        }
    }

    fn chars(&self) -> Chars<'a> {
        self.chars.clone()
    }

    fn chars_len(&self) -> usize {
        self.chars.as_str().len()
    }

    fn start(&self) -> usize {
        self.start
    }

    fn accept(&mut self, c: char) -> bool {
        match self.peek() {
            Some(peek_char) if peek_char == c => {
                self.next();
                true
            }
            _ => false,
        }
    }

    fn accept_while(&mut self, pred: impl Fn(char) -> bool) {
        while let Some(c) = self.peek() {
            if !pred(c) {
                break;
            }

            self.next();
        }
    }

    /// Returns the amount of bytes consumed by the lexer
    fn pos(&self) -> usize {
        self.input_len - self.chars_len()
    }

    fn peek(&self) -> Option<char> {
        self.chars().next()
    }

    fn peek_n(&self, n: usize) -> Option<char> {
        self.chars().nth(n)
    }

    fn bump_pos(&mut self) {
        self.start = self.pos()
    }

    fn text(&mut self) -> SmolStr {
        self.input[self.start()..self.pos()].into()
    }

    fn bump(&mut self) -> (Span, SmolStr) {
        let span = Span::new(self.start(), self.pos());
        let text = self.text();
        self.bump_pos();
        (span, text)
    }
}

impl Iterator for Source<'_> {
    type Item = char;

    fn next(&mut self) -> Option<Self::Item> {
        self.chars.next()
    }
}

#[derive(Debug)]
pub struct Lexer<'a> {
    source: Source<'a>,
}

impl<'a> Lexer<'a> {
    fn new(input: &'a str) -> Lexer<'a> {
        Lexer {
            source: Source::new(input),
        }
    }

    fn peek_n(&self) -> ParseResult<char> {
        self.source.peek().ok_or(ParseError::Eof)
    }

    fn next_token(&mut self) -> ParseResult<Token> {
        let next_char = self.source.next().eof()?;
        let peek_char = self.source.peek();

        let kind = match next_char {
            '/' => match peek_char {
                Some('/') => {
                    self.source.find(|c| *c == '\n');
                    self.source.bump_pos();
                    return self.next_token();
                }
                _ => {
                    self.source.next();
                    Tok![/]
                }
            },

            '=' => match peek_char {
                Some('=') => {
                    self.source.next();
                    Tok![==]
                }
                _ => Tok![=],
            },

            '!' => match peek_char {
                Some('=') => {
                    self.source.next();
                    Tok![!=]
                }
                _ => Tok![!],
            },

            '<' => match peek_char {
                Some('<') => {
                    self.source.next();
                    Tok![<<]
                }
                Some('=') => {
                    self.source.next();
                    Tok![<=]
                }
                _ => Tok![<],
            },

            '>' => match peek_char {
                Some('>') => {
                    self.source.next();
                    Tok![>>]
                }
                Some('=') => {
                    self.source.next();
                    Tok![>=]
                }
                _ => Tok![>],
            },

            ':' => match peek_char {
                Some('=') => {
                    self.source.next();
                    Tok![:=]
                }
                _ => {
                    Tok![:]
                }
            },

            ';' => Tok![;],
            ',' => Tok![,],
            '.' => Tok![.],

            '(' => Tok!['('],
            ')' => Tok![')'],
            '[' => Tok!['['],
            ']' => Tok![']'],
            '{' => Tok!['{'],
            '}' => Tok!['}'],

            // operators
            '*' => Tok![*],
            '%' => Tok![%],
            '+' => Tok![+],
            '-' => Tok![-],
            '~' => Tok![~],

            // literals
            '"' => return self.string_lit(),
            '0'..='9' => return self.number_lit(),

            // special
            _ if is_whitespace(next_char) => return self.whitespace(),
            _ if next_char.is_xid_start() => return self.maybe_ident(),
            _ => {
                self.source.bump_pos();
                return Err(ParseError::Unknown);
            }
        };

        let (span, text) = self.source.bump();

        Ok(Token { span, kind, text })
    }

    #[cfg(test)]
    fn lex_until_eof(&mut self) -> Vec<ParseResult<Token>> {
        let mut results = Vec::new();

        loop {
            let result = self.next_token();
            match result {
                Err(ParseError::Eof) => return results,
                t => results.push(t),
            }
        }
    }

    fn whitespace(&mut self) -> ParseResult<Token> {
        self.source.accept_while(is_whitespace);
        self.source.bump_pos();
        self.next_token()
    }

    fn string_lit(&mut self) -> ParseResult<Token> {
        self.source.find(|c| *c == '"').eof()?;
        let (span, text) = self.source.bump();
        Ok(Token {
            span,
            text,
            kind: Tok![string],
        })
    }

    fn number_lit(&mut self) -> ParseResult<Token> {
        self.source
            .accept_while(|c| ('0'..'9').contains(&c) || c == '_');
        let (span, text) = self.source.bump();
        Ok(Token {
            span,
            text,
            kind: Tok![number],
        })
    }

    fn maybe_ident(&mut self) -> ParseResult<Token> {
        self.source.accept_while(UnicodeXID::is_xid_continue);
        let (span, text) = self.source.bump();
        let kind = keyword(&text).unwrap_or(Tok![ident]);
        Ok(Token { span, text, kind })
    }
}

fn keyword(text: &str) -> Option<TokenKind> {
    Some(match text {
        "let" => Tok![let],
        "return" => Tok![return],
        "if" => Tok![if],
        "else" => Tok![else],
        "true" => Tok![true],
        "false" => Tok![false],
        _ => return None,
    })
}

/// True if `c` is considered a whitespace according to Rust language definition.
/// See [Rust language reference](https://doc.rust-lang.org/reference/whitespace.html)
/// for definitions of these classes.
pub fn is_whitespace(c: char) -> bool {
    // This is Pattern_White_Space.
    //
    // Note that this set is stable (ie, it doesn't change with different
    // Unicode versions), so it's ok to just hard-code the values.

    matches!(
        c,
        // Usual ASCII suspects
        '\u{0009}'   // \t
        | '\u{000A}' // \n
        | '\u{000B}' // vertical tab
        | '\u{000C}' // form feed
        | '\u{000D}' // \r
        | '\u{0020}' // space

        // NEXT LINE from latin1
        | '\u{0085}'

        // Bidi markers
        | '\u{200E}' // LEFT-TO-RIGHT MARK
        | '\u{200F}' // RIGHT-TO-LEFT MARK

        // Dedicated whitespace characters from Unicode
        | '\u{2028}' // LINE SEPARATOR
        | '\u{2029}' // PARAGRAPH SEPARATOR
    )
}

#[cfg(test)]
mod tests {
    use super::*;

    use std::fs;

    #[test]
    fn lexer() {
        insta::glob!("snapshot_inputs/lexer/*.txt", |path| {
            let input = fs::read_to_string(path).unwrap();
            let actual_tokens = Lexer::new(&input).lex_until_eof();
            let suffix = path.file_stem().unwrap().to_str().unwrap();
            insta::with_settings!({snapshot_path => "snapshots/lexer", snapshot_suffix => suffix}, {
                insta::assert_debug_snapshot!(actual_tokens)
            })
        })
    }

    #[test]
    fn pos() {
        let mut lexer = Lexer::new("test");
        assert_eq!(lexer.source.pos(), 0);
        assert!(lexer.source.next().is_some());
        assert_eq!(lexer.source.pos(), 1);
        assert!(lexer.source.next().is_some());
        assert_eq!(lexer.source.pos(), 2);
        assert!(lexer.source.next().is_some());
        assert_eq!(lexer.source.pos(), 3);
        assert!(lexer.source.next().is_some());
        assert_eq!(lexer.source.pos(), 4);
        assert!(lexer.source.next().is_none());
        assert_eq!(lexer.source.pos(), 4);
        assert!(lexer.source.next().is_none());
        assert_eq!(lexer.source.pos(), 4);
    }

    #[test]
    fn nothing() {
        assert!(Lexer::new("").lex_until_eof().is_empty());
    }

    #[test]
    fn whitespace() {
        assert!(Lexer::new("\t\t\n              \t\n")
            .lex_until_eof()
            .is_empty());
    }
}
