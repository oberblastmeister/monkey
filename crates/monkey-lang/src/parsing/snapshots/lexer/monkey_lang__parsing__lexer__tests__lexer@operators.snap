---
source: crates/monkey-lang/src/parsing/lexer.rs
expression: actual_tokens

---
[
    Ok(
        Token {
            span: Span {
                start: 0,
                end: 1,
            },
            kind: Asterisk,
            text: "*",
        },
    ),
    Ok(
        Token {
            span: Span {
                start: 2,
                end: 3,
            },
            kind: Modulo,
            text: "%",
        },
    ),
    Ok(
        Token {
            span: Span {
                start: 4,
                end: 5,
            },
            kind: Plus,
            text: "+",
        },
    ),
    Ok(
        Token {
            span: Span {
                start: 6,
                end: 7,
            },
            kind: Minus,
            text: "-",
        },
    ),
    Ok(
        Token {
            span: Span {
                start: 8,
                end: 9,
            },
            kind: Tilde,
            text: "~",
        },
    ),
    Ok(
        Token {
            span: Span {
                start: 10,
                end: 11,
            },
            kind: Gt,
            text: ">",
        },
    ),
    Ok(
        Token {
            span: Span {
                start: 12,
                end: 13,
            },
            kind: Lt,
            text: "<",
        },
    ),
    Ok(
        Token {
            span: Span {
                start: 14,
                end: 15,
            },
            kind: Bang,
            text: "!",
        },
    ),
    Ok(
        Token {
            span: Span {
                start: 16,
                end: 18,
            },
            kind: Slash,
            text: "/\n",
        },
    ),
]
